File: meetmind/database/loaders.py
"""文档加载器：把 JSON / Markdown / PDF / Word / 纯文本 文件统一解析为 RAG 文档列表。

每种文件格式对应一个独立函数，输入是 `Path`，输出是 `list[dict]`；
返回的 dict 至少包含 `content` 字段，可选 `type`、`source`。
顶层分发函数 `load_file()` 根据文件后缀决定调用哪个 loader。
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Callable

from meetmind.utils.logger import get_logger

logger = get_logger(__name__)


# 加载 JSON
def load_json(path: Path) -> list[dict]:
    """加载 JSON 文件。

    要求文件根是 `list[dict]`，每个 dict 至少有 `content` 字段，
    其余字段（type / date / source 等）会作为 metadata 一起保留。
    """
    raw = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(raw, list):
        logger.warning(f"JSON {path} 顶层不是 list，已跳过")
        return []
    docs: list[dict] = []

    # 拿到 list 后逐条检查，丢弃不符合要求的条目（非 dict 或缺 content）
    for i, item in enumerate(raw):
        if not isinstance(item, dict) or "content" not in item:
            continue
        item.setdefault("type", "json")
        item.setdefault("source", path.name)
        docs.append(item)
    return docs


# 加载 Markdown
def load_markdown(path: Path) -> list[dict]:
    """加载 Markdown 文件，按二级及以下标题或空行切块。

    简单策略：以 `##` 标题为分块边界；若没有标题，则按空行段落分块。
    每一块作为一条 RAG 文档。
    """
    text = path.read_text(encoding="utf-8")
    blocks: list[str] = []
    current: list[str] = []
    has_heading = False
    # 按行遍历，遇到 `##` 标题就切块；否则继续累积到 current
    for line in text.splitlines():
        if line.startswith("## "):
            has_heading = True
            if current:
                blocks.append("\n".join(current).strip())
            current = [line]
        else:
            current.append(line)
    if current:
        blocks.append("\n".join(current).strip())

    # 没有任何 `##` 标题 → 退化为按空行分段
    if not has_heading:
        blocks = []
        for b in text.split("\n\n"):
            blocks.append(b.strip())

    result = []
    for b in blocks:
        if b.strip():
            result.append({"content": b, "type": "markdown", "source": path.name})
    return result


# 加载 PDF
def load_pdf(path: Path) -> list[dict]:
    """加载 PDF 文件，每页作为一条 RAG 文档。

    使用 `pypdf` 提取文本。扫描版 PDF（图像）将得到空字符串并被丢弃。
    """
    from pypdf import PdfReader

    reader = PdfReader(str(path))
    docs: list[dict] = []
    # 把每页作为一个 doc，丢弃空页
    for i, page in enumerate(reader.pages, start=1):
        text = page.extract_text().strip()
        if not text:
            continue
        docs.append(
            {
                "content": text,
                "type": "pdf",
                "source": f"{path.name}#page{i}",
            }
        )
    return docs


# 加载 Word
def load_docx(path: Path) -> list[dict]:
    """加载 Word (.docx) 文件，按段落作为一条 RAG 文档。

    跳过空段落；表格内容暂不处理（如有需要可扩展遍历 `doc.tables`）。
    """
    from docx import Document

    document = Document(str(path))
    docs: list[dict] = []

    # 把每个段落作为一个 doc，丢弃空段落
    for para in document.paragraphs:
        text = para.text.strip()
        if not text:
            continue
        docs.append({"content": text, "type": "docx", "source": path.name})
    return docs


# 加载 纯文本
def load_text(path: Path) -> list[dict]:
    """加载 .txt 等纯文本文件，按空行分段。"""
    text : str= path.read_text(encoding="utf-8")
    raw_chunks : list[str] = text.split("\n\n")

    # 把切块后的每块文本去掉首尾空白，丢弃空块
    chunks = []
    for c in raw_chunks:
        stripped = c.strip()
        if stripped:
            chunks.append(stripped)

    result : list[dict] = []
    for c in chunks:
        result.append({"content": c, "type": "text", "source": path.name})
    return result


# 分发器
LOADERS: dict[str, Callable[[Path], list[dict]]] = {
    ".json": load_json,
    ".md": load_markdown,
    ".markdown": load_markdown,
    ".pdf": load_pdf,
    ".docx": load_docx,
    ".txt": load_text,
}


def load_file(path: Path) -> list[dict]:
    """根据文件后缀分发到对应 loader；返回类型为 list[dict]，每个 dict 至少有 `content` 字段。"""
    loader : Callable = LOADERS.get(path.suffix.lower())
    if loader is None:
        logger.warning(f"不支持的文件类型，跳过：{path.name}")
        return []

    try:
        return loader(path)
    except Exception as exc:
        logger.error(f"解析 {path} 失败：{exc}")
        return []

File: src/database/loaders.ts
/**
 * 文档加载器：把 JSON / Markdown / PDF / Word / 纯文本 文件统一解析为 RAG 文档列表。
 *
 * 每种文件格式对应一个独立函数，输入是文件绝对路径，输出是 RawDoc[]。
 * 顶层 `loadFile()` 根据扩展名分发。
 */

import { readFile } from "node:fs/promises";
import path from "node:path";

import { getLogger } from "../utils/logger.js";

const logger = getLogger("database.loaders");

export interface RawDoc {
  content: string;
  type?: string;
  date?: string;
  source?: string;
  // 允许 loader 写入额外字段（如 markdown 的 h1/h2 元信息），后续 splitter 会保留
  [extra: string]: unknown;
}

// ---------- JSON ----------

export async function loadJson(filePath: string): Promise<RawDoc[]> {
  const raw = await readFile(filePath, "utf-8");
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (exc) {
    logger.warning(`JSON ${filePath} 解析失败：${String(exc)}`);
    return [];
  }
  if (!Array.isArray(parsed)) {
    logger.warning(`JSON ${filePath} 顶层不是数组，已跳过`);
    return [];
  }
  const docs: RawDoc[] = [];
  for (const item of parsed) {
    if (item === null || typeof item !== "object") {
      continue;
    }
    const obj = item as Record<string, unknown>;
    if (typeof obj.content !== "string") {
      continue;
    }
    const doc: RawDoc = { content: obj.content };
    for (const [k, v] of Object.entries(obj)) {
      if (k === "content") {
        continue;
      }
      doc[k] = v;
    }
    if (doc.type === undefined) {
      doc.type = "json";
    }
    if (doc.source === undefined) {
      doc.source = path.basename(filePath);
    }
    docs.push(doc);
  }
  return docs;
}

// ---------- Markdown ----------

export async function loadMarkdown(filePath: string): Promise<RawDoc[]> {
  const text = await readFile(filePath, "utf-8");
  const baseName = path.basename(filePath);

  const lines = text.split(/\r?\n/);
  const blocks: string[] = [];
  let current: string[] = [];
  let hasHeading = false;

  for (const line of lines) {
    if (line.startsWith("## ")) {
      hasHeading = true;
      if (current.length > 0) {
        blocks.push(current.join("\n").trim());
      }
      current = [line];
    } else {
      current.push(line);
    }
  }
  if (current.length > 0) {
    blocks.push(current.join("\n").trim());
  }

  if (!hasHeading) {
    // 退化为按空行分段
    const newBlocks: string[] = [];
    for (const b of text.split("\n\n")) {
      const trimmed = b.trim();
      if (trimmed) {
        newBlocks.push(trimmed);
      }
    }
    blocks.splice(0, blocks.length, ...newBlocks);
  }

  const result: RawDoc[] = [];
  for (const b of blocks) {
    const stripped = b.trim();
    if (stripped) {
      result.push({ content: stripped, type: "markdown", source: baseName });
    }
  }
  return result;
}

// ---------- PDF ----------

export async function loadPdf(filePath: string): Promise<RawDoc[]> {
  // pdfjs-dist 4.x 默认入口就是 build/pdf.mjs（ESM）
  const pdfjs = await import("pdfjs-dist");
  const data = new Uint8Array(await readFile(filePath));
  const loadingTask = pdfjs.getDocument({
    data,
    // Node 没有 worker；显式关掉避免报错
    disableFontFace: true,
    isEvalSupported: false,
    useSystemFonts: false,
  });
  const pdf = await loadingTask.promise;

  const baseName = path.basename(filePath);
  const docs: RawDoc[] = [];
  for (let i = 1; i <= pdf.numPages; i++) {
    const page = await pdf.getPage(i);
    const tc = await page.getTextContent();
    const parts: string[] = [];
    for (const item of tc.items) {
      // pdfjs-dist 的 item 既可能是 TextItem 也可能是 TextMarkedContent；只关心前者
      const maybe = item as { str?: string };
      if (typeof maybe.str === "string" && maybe.str.length > 0) {
        parts.push(maybe.str);
      }
    }
    const text = parts.join(" ").trim();
    if (!text) {
      continue;
    }
    docs.push({
      content: text,
      type: "pdf",
      source: `${baseName}#page${i}`,
    });
  }
  return docs;
}

// ---------- DOCX ----------

export async function loadDocx(filePath: string): Promise<RawDoc[]> {
  const mammoth = await import("mammoth");
  const buffer = await readFile(filePath);
  const result = await mammoth.extractRawText({ buffer });
  const baseName = path.basename(filePath);

  // mammoth 输出按段落用 `\n` 分隔；过滤空段对齐 python-docx 的 doc.paragraphs 行为
  const docs: RawDoc[] = [];
  for (const para of result.value.split(/\r?\n/)) {
    const text = para.trim();
    if (!text) {
      continue;
    }
    docs.push({ content: text, type: "docx", source: baseName });
  }
  return docs;
}

// ---------- 纯文本 ----------

export async function loadText(filePath: string): Promise<RawDoc[]> {
  const text = await readFile(filePath, "utf-8");
  const baseName = path.basename(filePath);
  const result: RawDoc[] = [];
  for (const raw of text.split("\n\n")) {
    const stripped = raw.trim();
    if (stripped) {
      result.push({ content: stripped, type: "text", source: baseName });
    }
  }
  return result;
}

// ---------- 分发器 ----------

type LoaderFn = (filePath: string) => Promise<RawDoc[]>;

const LOADERS: Record<string, LoaderFn> = {
  ".json": loadJson,
  ".md": loadMarkdown,
  ".markdown": loadMarkdown,
  ".pdf": loadPdf,
  ".docx": loadDocx,
  ".txt": loadText,
};

/**
 * 根据文件扩展名分发到对应 loader。每个 RawDoc 至少有 `content`。
 */
export async function loadFile(filePath: string): Promise<RawDoc[]> {
  const ext = path.extname(filePath).toLowerCase();
  const loader = LOADERS[ext];
  if (!loader) {
    logger.warning(`不支持的文件类型，跳过：${path.basename(filePath)}`);
    return [];
  }
  try {
    return await loader(filePath);
  } catch (exc) {
    logger.error(`解析 ${filePath} 失败：${String(exc)}`);
    return [];
  }
}
